runtime/src/qio/qio.c

```diff
@@ -2829,26 +2829,34 @@ qioerr _qio_buffered_read(qio_channel_t* ch, void* ptr, ssize_t len, ssize_t* am
   ) {
     // copy out what remains in the buffer before making a system call
     gotlen = qio_ptr_diff(ch->cached_end, ch->cached_cur);
-    if ( gotlen > 0 ) {
+
+    if ( ch->cached_start == NULL ) {
+      // only use the right-mark-start offset if the buffered offset does not exist
       start = qbuffer_iter_at(&ch->buf, qio_channel_offset_unlocked(ch));
-      // start = qbuffer_iter_at(&ch->buf, ch->cached_start_pos + qio_ptr_diff(ch->cached_cur, ch->cached_start));
-      end = start;
-      qbuffer_iter_advance(&ch->buf, &end, gotlen);  // end of available data
+    } else {
+      start = qbuffer_iter_at(&ch->buf, ch->cached_start_pos + qio_ptr_diff(ch->cached_cur, ch->cached_start));
+    }
+    end = start;
+    qbuffer_iter_advance(&ch->buf, &end, gotlen);  // end of available data
 
+    if ( gotlen > 0 ) {
       // copy 'gotlen' bytes into the ptr
       err = qbuffer_copyout(&ch->buf, start, end, ptr, gotlen);
       if( err ) return err;
+    }
 
-      // advance the ptr, start of available data, etc.
-      ptr = qio_ptr_add(ptr, gotlen);
-      _set_right_mark_start(ch, end.offset);
-      ch->cached_cur = qio_ptr_add(ch->cached_cur, gotlen);
-      remaining -= gotlen;
-      *amt_read = gotlen;
+    // advance the ptr, start of available data, etc.
+    ptr = qio_ptr_add(ptr, gotlen);
+    _set_right_mark_start(ch, end.offset);
+    ch->cached_cur = qio_ptr_add(ch->cached_cur, gotlen);
+    remaining -= gotlen;
+    *amt_read = gotlen;
 
-      // clean up the now unused portion of the buffer
-      qbuffer_trim_front(&ch->buf, gotlen);
-    }
+    // clean up the now unused portion of the buffer
+    qbuffer_trim_front(&ch->buf, gotlen);
+    ch->cached_cur = NULL;
+    ch->cached_end = NULL;
+    ch->cached_start = NULL;
```

Comment:
Why is `cache_cur` set on line 2851 only to be cleared out here?

Reply:
Setting `cache_cur` there is unneeded. I've removed it.